File: csharp/Token.cs
namespace WildlingLib;

public sealed class TokenOptions
{
    public string? String { get; set; }
    public int? StartLength { get; set; }
    public int? EndLength { get; set; }
    public List<string>? Variants { get; set; }
    public string Src { get; set; } = "";

    public static TokenOptions Of(List<string> variants, int startLength, int endLength, string src) =>
        new()
        {
            Variants = variants,
            StartLength = startLength,
            EndLength = endLength,
            Src = src,
        };
}

public sealed class Token
{
    private readonly string _src;
    private readonly int _startLength;
    private readonly int _endLength;
    private readonly List<string> _variants;
    private readonly int _count;

    public Token(TokenOptions options)
    {
        _src = options.Src ?? "";
        _startLength = DefaultInteger(options.StartLength, 1);
        _endLength = DefaultInteger(options.EndLength, 1);
        _variants = options.Variants != null ? new List<string>(options.Variants) : new List<string>();

        var total = 0;
        for (var length = _startLength; length <= _endLength; length++)
        {
            total += Pow(_variants.Count, length);
        }
        _count = total;
    }

    private static int DefaultInteger(int? option, int fallback) =>
        option is >= 0 ? option.Value : fallback;

    private static int Pow(int baseValue, int exp)
    {
        var result = 1;
        for (var i = 0; i < exp; i++)
        {
            result *= baseValue;
        }
        return result;
    }

    public int Count() => _count;

    public string Src() => _src;

    public string Get(int index)
    {
        if (index > _count - 1 || index < 0)
        {
            return "";
        }

        if (index == 0 && _startLength == 0)
        {
            return "";
        }

        var indexWithOffset = index;
        var stringLength = _startLength;
        for (stringLength = _startLength; stringLength <= _endLength; stringLength++)
        {
            var offsetCount = Pow(_variants.Count, stringLength);
            if (indexWithOffset < offsetCount)
            {
                break;
            }
            indexWithOffset -= offsetCount;
        }

        var chars = new string[stringLength];
        for (var i = 0; i < stringLength; i++)
        {
            var variantIndex = indexWithOffset % _variants.Count;
            indexWithOffset /= _variants.Count;
            chars[i] = _variants[variantIndex];
        }
        return string.Concat(chars);
    }
}

File: fsharp/Token.fs
namespace WildlingLib

open System.Collections.Generic

type TokenOptions =
    { mutable String: string option
      mutable StartLength: int option
      mutable EndLength: int option
      mutable Variants: ResizeArray<string> option
      mutable Src: string }

module TokenOptions =
    let ofVariants (variants: ResizeArray<string>) startLength endLength src =
        { String = None
          StartLength = Some startLength
          EndLength = Some endLength
          Variants = Some variants
          Src = src }

type Token(options: TokenOptions) =
    let src = if isNull options.Src then "" else options.Src

    let defaultInteger (option: int option) fallback =
        match option with
        | Some v when v >= 0 -> v
        | _ -> fallback

    let startLength = defaultInteger options.StartLength 1
    let endLength = defaultInteger options.EndLength 1

    let variants =
        match options.Variants with
        | Some v -> ResizeArray<string>(v)
        | None -> ResizeArray<string>()

    let pow baseValue exp =
        let mutable result = 1
        for _ = 1 to exp do
            result <- result * baseValue
        result

    let count =
        let mutable total = 0
        for length = startLength to endLength do
            total <- total + pow variants.Count length
        total

    member _.Count() = count
    member _.Src() = src

    member _.Get(index: int) =
        if index > count - 1 || index < 0 then
            ""
        elif index = 0 && startLength = 0 then
            ""
        else
            let mutable indexWithOffset = index
            let mutable stringLength = startLength
            let mutable found = false
            let mutable length = startLength
            while length <= endLength && not found do
                let offsetCount = pow variants.Count length
                if indexWithOffset < offsetCount then
                    stringLength <- length
                    found <- true
                else
                    indexWithOffset <- indexWithOffset - offsetCount
                    length <- length + 1

            let chars = Array.zeroCreate<string> stringLength
            for i = 0 to stringLength - 1 do
                let variantIndex = indexWithOffset % variants.Count
                indexWithOffset <- indexWithOffset / variants.Count
                chars.[i] <- variants.[variantIndex]
            String.concat "" chars
